src/View/Helper/PaginatorHelper.php

```diff
@@ -635,96 +635,180 @@ public function numbers(array $options = [])
             $templater->{$method}($options['templates']);
         }
 
+        if ($options['modulus'] && $params['pageCount'] > $options['modulus']) {
+            $out = $this->_modulusNumbers($templater, $params, $options);
+        } else {
+            $out = $this->_numbers($templater, $params, $options);
+        }
+
+        if (isset($options['templates'])) {
+            $templater->pop();
+        }
+
+        return $out;
+    }
+
+    /**
+     * Calculates the start and end for the pagination numbers.
+     *
+     * @param array $params Params from the numbers() method.
+     * @param array $options Options from the numbers() method.
+     * @return array An array with the start and end numbers.
+     */
+    protected function _getNumbersStartAndEnd($params, $options)
+    {
+        $half = (int)($options['modulus'] / 2);
+        $end = $params['page'] + $half;
+
+        if ($end > $params['pageCount']) {
+            $end = $params['pageCount'];
+        }
+        $start = $params['page'] - ($options['modulus'] - ($end - $params['page']));
+        if ($start <= 1) {
+            $start = 1;
+            $end = $params['page'] + ($options['modulus'] - $params['page']) + 1;
+        }
+        return compact('start', 'end');
+    }
+
+    /**
+     * Formats a number for the paginator number output.
+     *
+     * @param StringTemplate $templater
+     * @param array $options Options from the numbers() method.
+     * @return string
+     */
+    protected function _formatNumber($templater, $options)
+    {
+        $vars = [
+            'text' => $options['text'],
+            'url' => $this->generateUrl(['page' => $options['page']], $options['model']),
+        ];
+        return $templater->format('number', $vars);
+    }
+
+    /**
+     * Generates the numbers for the paginator numbers() method.
+     *
+     * @param StringTemplate $templater
+     * @param array $params Params from the numbers() method.
+     * @param array $options Options from the numbers() method.
+     * @return string Markup output.
+     */
+    protected function _modulusNumbers($templater, $params, $options)
+    {
         $out = '';
         $ellipsis = $templater->format('ellipsis', []);
 
-        if ($options['modulus'] && $params['pageCount'] > $options['modulus']) {
-            $half = (int)($options['modulus'] / 2);
-            $end = $params['page'] + $half;
+        extract($this->_getNumbersStartAndEnd($params, $options));
 
```

Comment:
It never caused me trouble, would you mind to explain what kind of badness can happen with extract()? Thank you.